test/multi_rollup_sequencer.go

```diff
@@ -0,0 +1,134 @@
+	lastBatchHashMutex sync.RWMutex
+
+	seenBatches      map[string]struct{}
+	seenBatchesMutex sync.Mutex
+}
+
+// SubmitRollupTransaction implements sequencing.Sequencer.
+func (d *MultiRollupSequencer) SubmitRollupTransaction(ctx context.Context, req sequencing.SubmitRollupTransactionRequest) (*sequencing.SubmitRollupTransactionResponse, error) {
+	rollup, err := d.getOrCreateRollup(req.RollupId)
+	if err != nil {
+		return nil, err
+	}
+	rollup.tq.AddTransaction(req.Tx)
+	return &sequencing.SubmitRollupTransactionResponse{}, nil
+}
+
+// GetNextBatch implements sequencing.Sequencer.
+func (d *MultiRollupSequencer) GetNextBatch(ctx context.Context, req sequencing.GetNextBatchRequest) (*sequencing.GetNextBatchResponse, error) {
+	rollup, err := d.getOrCreateRollup(req.RollupId)
+	if err != nil {
+		return nil, err
+	}
+
+	now := time.Now()
+	rollup.lastBatchHashMutex.RLock()
+	lastBatchHash := rollup.lastBatchHash
+	rollup.lastBatchHashMutex.RUnlock()
+
+	if lastBatchHash == nil && req.LastBatchHash != nil {
+		return nil, errors.New("lastBatch is supposed to be nil")
+	} else if lastBatchHash != nil && req.LastBatchHash == nil {
+		return nil, errors.New("lastBatch is not supposed to be nil")
+	} else if !bytes.Equal(lastBatchHash, req.LastBatchHash) {
+		return nil, errors.New("supplied lastBatch does not match with sequencer last batch")
+	}
+
+	batch := rollup.tq.GetNextBatch(req.MaxBytes)
+	batchRes := &sequencing.GetNextBatchResponse{Batch: batch, Timestamp: now}
+	// If there are no transactions, return empty batch without updating the last batch hash
+	if len(batch.Transactions) == 0 {
+		return batchRes, nil
+	}
+
+	h, err := batch.Hash()
+	if err != nil {
+		return nil, err
+	}
+
+	rollup.lastBatchHashMutex.Lock()
+	rollup.lastBatchHash = h
+	rollup.lastBatchHashMutex.Unlock()
+
+	rollup.seenBatchesMutex.Lock()
+	rollup.seenBatches[hex.EncodeToString(h)] = struct{}{}
+	rollup.seenBatchesMutex.Unlock()
+	return batchRes, nil
+}
+
+// VerifyBatch implements sequencing.Sequencer.
+func (d *MultiRollupSequencer) VerifyBatch(ctx context.Context, req sequencing.VerifyBatchRequest) (*sequencing.VerifyBatchResponse, error) {
+	rollup, err := d.getOrCreateRollup(req.RollupId)
+	if err != nil {
+		return nil, err
+	}
+
+	rollup.seenBatchesMutex.Lock()
+	defer rollup.seenBatchesMutex.Unlock()
+	key := hex.EncodeToString(req.BatchHash)
+	if _, exists := rollup.seenBatches[key]; exists {
+		return &sequencing.VerifyBatchResponse{Status: true}, nil
+	}
+	return &sequencing.VerifyBatchResponse{Status: false}, nil
+}
+
+// getOrCreateRollup returns the RollupData for a given rollupId, creating it if necessary.
+func (d *MultiRollupSequencer) getOrCreateRollup(rollupId []byte) (*RollupData, error) {
+	rollupKey := hex.EncodeToString(rollupId)
+
+	d.rollupsMutex.RLock()
+	rollup, exists := d.rollups[rollupKey]
+	d.rollupsMutex.RUnlock()
+
+	if exists {
+		return rollup, nil
+	}
+
+	d.rollupsMutex.Lock()
+	defer d.rollupsMutex.Unlock()
```

Comment:
I'm worries that in testing, due to the rapid cycles, that this could cause flakiness in tests if they are constructed well. Specifically in understanding how the `MultiRollupSequencer` is going to be used within tests or across tests. 

I think this would be a safer way to handle the mutex.
```suggestion
	d.rollupsMutex.Lock()
	defer d.rollupsMutex.Unlock()
	rollup, exists := d.rollups[rollupKey]
	if exists {
		return rollup, nil
	}
```